File: backend/app/schemas/reports.py
from datetime import datetime
from typing import Any, Dict, List, Optional
from pydantic import BaseModel


class ReportSummary(BaseModel):
    total_events: int
    honeypot_hits_24h: int
    blocked_ips_count: int
    active_alerts_count: int
    mtd_rotations_count: int
    total_audit_logs: int
    camera_violations_count: int
    last_generated: Optional[str] = None


class ReportItem(BaseModel):
    id: str
    title: str
    type: str  # "Security Audit", "Incident Log", "Camera Performance", "MTD Analytics", "System Audit Log"
    generatedAt: str
    size: str
    format: str  # "PDF", "CSV", "JSON"
    description: str
    download_url: Optional[str] = None


class ReportGenerateRequest(BaseModel):
    report_type: str  # "security_audit", "incident_log", "camera_performance", "mtd_analytics", "audit_log"
    format: str = "JSON"  # "JSON", "CSV", "PDF"
    date_range: Optional[str] = "24h"  # "24h", "7d", "30d", "all"


class ReportGenerateResponse(BaseModel):
    id: str
    title: str
    type: str
    generatedAt: str
    size: str
    format: str
    description: str
    summary: Dict[str, Any]
    data: List[Dict[str, Any]]


class AuditLogOut(BaseModel):
    id: int
    user_email: str
    action: str
    resource: str
    details: Optional[str] = None
    ip_address: Optional[str] = None
    timestamp: datetime

    class Config:
        from_attributes = True

File: frontend/src/api/reports.ts
import api from "./client";

export interface ReportSummary {
  total_events: number;
  honeypot_hits_24h: number;
  blocked_ips_count: number;
  active_alerts_count: number;
  mtd_rotations_count: number;
  total_audit_logs: number;
  camera_violations_count: number;
  last_generated?: string;
}

export interface ReportItem {
  id: string;
  title: string;
  type: string;
  generatedAt: string;
  size: string;
  format: string;
  description: string;
  download_url?: string;
}

export interface ReportGenerateRequest {
  report_type: string;
  format?: string;
  date_range?: string;
}

export interface ReportGenerateResponse {
  id: string;
  title: string;
  type: string;
  generatedAt: string;
  size: string;
  format: string;
  description: string;
  summary: Record<string, any>;
  data: Array<Record<string, any>>;
}

export interface AuditLogOut {
  id: number;
  user_email: string;
  action: string;
  resource: string;
  details?: string;
  ip_address?: string;
  timestamp: string;
}

export const getReports = async (): Promise<ReportItem[]> => {
  const response = await api.get("/api/v1/reports");
  return response.data;
};

export const getReportSummary = async (): Promise<ReportSummary> => {
  const response = await api.get("/api/v1/reports/summary");
  return response.data;
};

export const generateReport = async (
  request: ReportGenerateRequest
): Promise<ReportGenerateResponse> => {
  const response = await api.post("/api/v1/reports/generate", request);
  return response.data;
};

export const getAuditLogs = async (): Promise<AuditLogOut[]> => {
  const response = await api.get("/api/v1/reports/audit-logs");
  return response.data;
};
